lib/srfi/35/internal.scm

```diff
@@ -0,0 +1,206 @@
+(define-record-type Simple-Condition
+  (make-simple-condition)
+  simple-condition?)
+
+(define-record-type Compound-Condition
```

Comment:
On reviewing the SRFI I'm rather unhappy with it because of this distinction.  It's ad-hoc and leads to a lot of ambiguity which doesn't exist in other exception systems.  Cleaner is to always have one type which may have fields referring to other types (e.g. the underlying causes).  I guess this was meant to make up for the lack of multiple inheritance.

Reply:
Sorry, I wasn't clear in my complaint.  The complaint is in the existence of, and design of, compound conditions, not any implementation details.

The fundamental problem is that in SRFI 35 a given condition doesn't have any single, or even most specific, type.  You move the burden of taking apart and understanding the condition to the call site, which isn't in general prepared or qualified to deal with this.  Instead the code raising the condition should take care to choose a single most appropriate type.  If there is a root cause it can be attached as a field, so that where it really matters the caller can unwrap and understand the condition in more detail.